File: Backend/src/main/java/com/gestion_actas/gestion/de/actas/components/JwtUtil.java
package com.gestion_actas.gestion.de.actas.components;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.util.Date;

@Component
public class JwtUtil {

    private final SecretKey key;
    private final long expiration;

    public JwtUtil(
            @Value("${jwt.secret}") String secretBase64,
            @Value("${jwt.expiration:86400000}") long expiration
    ) {
        this.key = Keys.hmacShaKeyFor(Decoders.BASE64.decode(secretBase64));
        this.expiration = expiration;
    }

    public String generateToken(String username, String rol) {
        Date now = new Date();
        Date exp = new Date(now.getTime() + expiration);

        return Jwts.builder()
                .subject(username)
                .claim("rol", rol)
                .issuedAt(now)
                .expiration(exp)
                .signWith(key) // o .signWith(key, Jwts.SIG.HS256)
                .compact();
    }

    public String extractUsername(String token) {
        return claims(token).getSubject();
    }

    public String extractRol(String token) {
        return claims(token).get("rol", String.class);
    }

    public boolean validateToken(String token) {
        try {
            Jwts.parser()
                    .verifyWith(key)
                    .build()
                    .parseSignedClaims(stripBearer(token));
            return true;
        } catch (JwtException | IllegalArgumentException e) {
            return false;
        }
    }

    private Claims claims(String token) {
        return Jwts.parser()
                .verifyWith(key)
                .build()
                .parseSignedClaims(stripBearer(token))
                .getPayload();
    }

    private String stripBearer(String token) {
        if (token != null && token.startsWith("Bearer ")) {
            return token.substring(7);
        }
        return token;
    }
}

File: BackendActasPHP/app/Utils/JwtUtil.php
<?php

namespace App\Helpers;

use Firebase\JWT\JWT;
use Firebase\JWT\Key;

class JwtUtil
{
    private $key;
    private $expiration;

    public function __construct()
    {
        $this->key = env('JWT_SECRET');
        $this->expiration = env('JWT_EXPIRATION', 86400);
    }

    public function generateToken($username, $rol)
    {
        $now = time();
        $exp = $now + $this->expiration;

        $payload = [
            'sub' => $username,
            'rol' => $rol,
            'iat' => $now,
            'exp' => $exp,
        ];

        return JWT::encode($payload, $this->key, 'HS256');
    }

    public function extractUsername($token)
    {
        $payload = $this->claims($token);
        return $payload['sub'] ?? null;
    }

    public function extractRol($token)
    {
        $payload = $this->claims($token);
        return $payload['rol'] ?? null;
    }

    public function validateToken($token)
    {
        try {
            JWT::decode($token, new Key($this->key, 'HS256'));
            return true;
        } catch (\Exception $e) {
            return false;
        }
    }

    public function claims($token)
    {
        return (array) JWT::decode($token, new Key($this->key, 'HS256'));
    }
}